src/geoFire/geoQuery.ts

```diff
@@ -0,0 +1,526 @@
+    // geohashes were queried even if the client updates the radius of the query.
+    // This results in no 'READY' event being fired after the .updateCriteria() call.
+    // Check to see if this is the case, and trigger the 'READY' event.
+    if (geohashesToQuery.length === 0) {
+      this._geohashQueryReadyCallback();
+    }
+  }
+
+  /**
+   * Encodes a query as a string for easier indexing and equality.
+   *
+   * @param query The query to encode.
+   * @returns The encoded query as string.
+   */
+  private _queryToString(query: string[]): string {
+    if (query.length !== 2) {
+      throw new Error('Not a valid geohash query: ' + query);
+    }
+    return query[0] + ':' + query[1];
+  }
+
+  /**
+   * Removes the location from the local state and fires any events if necessary.
+   *
+   * @param key The key to be removed.
+   * @param currentLocation The current location as [latitude, longitude] pair or null if removed.
+   */
+  private _removeLocation(key: string, currentLocation?: number[]): void {
+    const locationDict = this._locationsTracked[key];
+    delete this._locationsTracked[key];
+    if (typeof locationDict !== 'undefined' && locationDict.isInQuery) {
+      const distanceFromCenter: number = (currentLocation) ? distance(currentLocation, this._center) : null;
+      this._fireCallbacksForKey('key_exited', key, currentLocation, distanceFromCenter);
+    }
+  }
+
+  /**
+   * Decodes a query string to a query
+   *
+   * @param str The encoded query.
+   * @returns The decoded query as a [start, end] pair.
+   */
+  private _stringToQuery(str: string): string[] {
+    const decoded: string[] = str.split(':');
+    if (decoded.length !== 2) {
+      throw new Error('Invalid internal state! Not a valid geohash query: ' + str);
+    }
+    return decoded;
+  }
+
+  /**
+   * Callback for any updates to locations. Will update the information about a key and fire any necessary
+   * events every time the key's location changes.
+   *
+   * When a key is removed from GeoFire or the query, this function will be called with null and performs
+   * any necessary cleanup.
+   *
+   * @param key The key of the geofire location.
+   * @param location The location as [latitude, longitude] pair.
+   */
+  private _updateLocation(key: string, location?: number[]): void {
+    validateLocation(location);
+    // Get the key and location
+    let distanceFromCenter: number, isInQuery;
+    const wasInQuery: boolean = (key in this._locationsTracked) ? this._locationsTracked[key].isInQuery : false;
+    const oldLocation: number[] = (key in this._locationsTracked) ? this._locationsTracked[key].location : null;
+
+    // Determine if the location is within this query
+    distanceFromCenter = distance(location, this._center);
+    isInQuery = (distanceFromCenter <= this._radius);
+
+    // Add this location to the locations queried dictionary even if it is not within this query
+    this._locationsTracked[key] = {
```

Comment:
Similar to above, we should provide an interface for this type, and we can remove the duplication of identical key/val here.